sympy/physics/quantum/qapply.py

```diff
@@ -1,206 +1,364 @@
+# supported in general.
+
+@lapply_mul2types_B.register(QExpr, (QExpr, Expr))
+def hdlrQQ_for(lhs:QExpr, rhs:Expr, **options) -> Union[Expr, List, None]:
+    # The quantum package uses methods _apply_operator and _apply_from_right
+    # methods for many of its objects to describe application. So for quantum
+    # objects try to apply _apply_operator and _apply_from_right_to as a
+    # catch-all. multipledispatch will give priority to more specific rules.
+    res = cast(Any, None)
+    if hasattr(lhs, '_apply_operator'):
+        try: # for historical reasons, _apply_operator is tried first
+            res = cast(Any, lhs._apply_operator(rhs, **options)) # type: ignore[attr-defined]
+        except (NotImplementedError): # action lhs not implemented for rhs
+            pass
+    if res is None and isinstance(rhs, QExpr):
+        # Invoke the catch-all function for Expr*QExpr
+        res = (lapply_mul2types_B.dispatch(Expr, QExpr))(lhs, rhs, **options)
+    if res is None and options["dagger"]:
+        res = lapply_mul2daggered(lhs, rhs, **options)
+    return res # either some result or None # type: ignore[no-redef]
+
+@lapply_mul2types_B.register(Expr, QExpr) # sides (QExpr, Expr) switched
+def hdlrEQ_for(lhs:Expr, rhs:QExpr, **options) -> Union[Expr, List, None]:
+    res = cast(Any, None)
+    if hasattr(rhs, '_apply_from_right_to'):
+        try:
+            res = cast(Any, rhs._apply_from_right_to(lhs, **options)) # type: ignore[attr-defined]
+        except (NotImplementedError): # action rhs not implemented for lhs
+            pass
+    if res is None and options["dagger"]:
+        res = lapply_mul2daggered(lhs, rhs, **options)
+    return res # either some result or None # type: ignore[no-redef]
+
+# If Layer A had no result for QExpr, QExpr, and options["dagger"]==True,
+# try identity lhsR*rhsL = Dagger(Dagger(lhsR*rhsL))
+#                        = Dagger(Dagger(rhsL)*Dagger(lhsR))
+# as the last resort. Returns Mul(lhs, rhs, evaluate=False) if no success.
+# Dagger is built on class functions.elementary.complexes.adjoint, which
+# is also used in matrices.Adjoint, so it is actually a general class and
+# independent of package quantum where it is located! However the option
+# "dagger" makes sense only in packages like quantum which has one-sided
+# definition of 'apply' resp. *, i.e. where A*B may be undefined, but
+# Dagger(B)*Dagger(A) is defined.
+
+def lapply_mul2daggered(lhs:Expr, rhs:Expr, **options) \
+                                                 -> Union[Expr, None]:
+    #assert options["dagger"] # precondition for this function
+    res0 = lhs * rhs # try the standard last resort first
+    if lapply_no_act(lhs, rhs, res0):
+        res0 = Mul(lhs, rhs, evaluate=False)
+    else: return res0
+    # Return if the daggered values are just formal adjoints
+    if isinstance((rd := Dagger(rhs)), adjoint): return res0
+    if isinstance((ld := Dagger(lhs)), adjoint): return res0

```

Comment:
Most uses of the Walrus operator here seem to obfuscate the code. The same goes for inlining of conditionals. The idea of Python's significant whitespace is that endentation is supposed to render the logical structure of the code visually so that you can clearly delineate the different blocks.

Something like this would be clearer if written like:
```python
rd = Dagger(rhs)
if isinstance(rd, adjoint):
    return res0

ld = Dagger(lhs)
if isinstance(ld, adjoint):
    return res0
```
With this it is easy to see the logic both ways. There are two branches that go into the condition block and return. There is another branch that passes over the two blocks and in that case both `ld` and `rd` will have been assigned for the benefit of subsequent code.

Or since `res0` is returned either way:
```python
ld = Dagger(lhs)
rd = Dagger(rhs)
if isinstance(ld, adjoint) or isinstance(rd, adjoint):
    return res0
```
Or perhaps we could use more descriptive variable names:
```python
lhs_dagger = Dagger(lhs)
rhs_dagger = Dagger(rhs)

# Return if the daggered values are just formal adjoints
if isinstance(lhs_dagger, adjoint) or isinstance(rhs_dagger, adjoint):
    return mul_lhs_rhs
```

Reply:
Concerning amount of comments - I would have liked to write all comments in columns 81-132. Everybody has wide screens today. Code would not be cluttered with comments, but still comments would be at the code...

Reply:
Concerning the structure of the code:  

The explanations from Jan 16th on still hold (replace "atomic" by "elementary" and "c_nc_atom" by "c_nc_ncef"), but the separation of object specific knowledge and the object agnostic application algorithm in lapply_mul2(), lapply_c_nc_ncef() has now been done.

I admit that the algorithm is somewhat cluttered by the code für the cache for Powers, so it's probably best to ignore that for the moment.

It's also useful to see c_nc_ncef and c_nc_ncef_powi as black boxes for the moment: This is almost the majority of code, and but all they do is looking for special options to simplify powers (and prepare the cache for power bases). 

So the core algorithm in lapply_mul2 and lapply_c_nc_ncef does nothing more than:
- split the expression into factors. Pull up commutative factors to the front; they are dealt with separately.
- recursively dig into the right most nc factor until an elementary factor `rhs` has been found
- get another nc elementary factor `lhs` to the left
- apply the two to each other using multipledispatch handlers
- If successful (i.e. a new object emerged, not just `Mul(lhs, rhs)`), multiply that to factors to the right (if any), and continue from there
- If not successful, leave rhs, get a new factor to the left, and continue from there.
- if we hit an Add(..), distribute the factors over summands.

Commutative factors are expanded and simplified by a subset of expand() that is not expanding into function arguments as expand(deep=True) does; they are also mixed with the nc-factors to permit annulation of summands.